File: carbondesign/tags/grid.py
"""
Grid
====

See: https://the-carbon-components.netlify.app/?nav=grid

""" # pylint:disable=line-too-long
# pylint:disable=too-many-lines

from .base import Node

class Grid(Node):
    """Grid wrapper component.
    """
    WANT_CHILDREN = True
    "Template Tag needs closing end tag."
    MODES = ('default', 'bleed')
    "Available variants."
    NODE_PROPS = ('full_width', 'gap')
    "Extended Template Tag arguments."
    GAP_SIZES = ('narrow', 'condensed')
    "Grid gap sizes."

    def prepare(self, values, context):
        """Prepare values for rendering the templates.
        """
        if self.eval(self.kwargs.get('full_width'), context):
            values['class'].append('bx--grid--full-width')

        gap = self.eval(self.kwargs.get('gap'), context)
        if gap in self.GAP_SIZES:
            values['class'].append(f'bx--grid--{gap}')


    def render_default(self, values, context):
        """Output html of the component.
        """
        template = '<{astag} class="bx--grid {class}" {props}>{child}</{astag}>'
        return self.format(template, values)


    def render_bleed(self, values, context):
        """Output html of the component.
        """
        template = """
<div class="bleed">
  <{astag} class="bx--grid {class}" {props}>{child}</{astag}>
</div>
"""
        return self.format(template, values)


class Row(Node):
    """Grid Row component.
    """
    WANT_CHILDREN = True
    "Template Tag needs closing end tag."

    def render_default(self, values, context):
        """Output html of the component.
        """
        template = '<{astag} class="bx--row {class}" {props}>{child}</{astag}>'
        return self.format(template, values)


class Column(Node):
    """Grid columns.
    """
    WANT_CHILDREN = True
    "Template Tag needs closing end tag."
    COL_SIZES = ('sm', 'md', 'lg', 'xlg', 'max')
    "Column sizes."
    NODE_PROPS = (*COL_SIZES, *[f'offset_{x}' for x in COL_SIZES])
    "Extended Template Tag arguments."

    has_size = False

    def prepare(self, values, context):
        """Prepare values for rendering the templates.
        """
        for size in self.COL_SIZES:
            width = self.eval(self.kwargs.get(size), context)
            if width:
                self.has_size = True
                values['class'].append(f'bx--col-{size}-{width}')

            width = self.eval(self.kwargs.get(f'offset_{size}'), context)
            if width:
                values['class'].append(f'bx--offset-{size}-{width}')
        if not self.has_size:
            values['class'].append('bx--col bx--col--auto')


    def render_default(self, values, context):
        """Output html of the component.
        """
        if self.has_size:
            template = """
<{astag} class="{class}" {props}>
  <div class="outside">
    <div class="inside">
      {child}
    </div>
  </div>
</{astag}>
"""
        else:
            template = '<{astag} class="{class}" {props}>{child}</{astag}>'
        return self.format(template, values)


class AspectRatio(Node):
    """Grid aspect ratio.
    """
    WANT_CHILDREN = True
    "Template Tag needs closing end tag."
    NODE_PROPS = ('ratio',)
    "Extended Template Tag arguments."

    def prepare(self, values, context):
        """Prepare values for rendering the templates.
        """
        ratio = self.eval(self.kwargs.get('ratio'), context)
        if ratio:
            values['class'].append(f'bx--aspect-ratio--{ratio}')


    def render_default(self, values, context):
        """Output html of the component.
        """
        template = """
<{astag} class="bx--aspect-ratio {class}" {props}>
  <div class="bx--aspect-ratio--object">
    {child}
  </div>
</{astag}>
"""
        return self.format(template, values)


components = {
    'Grid': Grid,
    'Row': Row,
    'Col': Column,
    'AspectRatio': AspectRatio,
}

File: carbondesign/mithril-js/grid.js
import m from 'mithril/hyperscript';
//-
import { Node, modify_svg } from './base';

export class Grid extends Node
{
    WANT_CHILDREN = true
    MODES = ['default', 'bleed']
    NODE_PROPS = ['full_width', 'gap']
    GAP_SIZES = ['narrow', 'condensed']

    prepare(vnode, values, context)
    {
        if (vnode.attrs.full_width)
        {
            values['class'].push('bx--grid--full-width');
        }

        let gap = vnode.attrs.gap;
        if (gap)
        {
            values['class'].push(`bx--grid--${gap}`);
        }
    }

    render_default(vnode, values, context)
    {
        return (
//##
m(values.tag,
  {
    'class': `bx--grid ${values['class']}`,
    ...values.props,
  },
  values.child)
//##
        );
    }

    render_bleed(vnode, values, context)
    {
        return (
//##
m('div.bleed', null,
  m(values.tag,
    {
      'class': `bx--grid ${values['class']}`,
      ...values.props,
    },
    values.child))
//##
        );
    }
}


export class Row extends Node
{
    WANT_CHILDREN = true

    render_default(vnode, values, context)
    {
        return (
//##
m(values.tag,
  {
    'class': `bx--row ${values['class']}`,
    ...values.props,
  },
  values.child)
//##
        );
    }
}


export class Col extends Node
{
    WANT_CHILDREN = true
    COL_SIZES = ['sm', 'md', 'lg', 'xlg', 'max']
    NODE_PROPS = COL_SIZES + COL_SIZES.map(x => `offset_${x}`)

    prepare(vnode, values, context)
    {
        for (let size of this.COL_SIZES)
        {
            let width = vnode.attrs[size]
            if (width)
            {
                this.has_size = true;
                values['class'].push(`bx--col-${size}-${width}`);
            }

            width = vnode.attrs[`offset-${size}`];
            if (width)
            {
                values['class'].push(`bx--offset-${size}-${width}`);
            }
        }
        if (!this.has_size)
        {
            values['class'].push('bx--col');
            values['class'].push('bx--col--auto');
        }
    }

    render_default(vnode, values, context)
    {
        if (this.has_size)
        {
            return (
//##
m(values.tag,
  {
    'class': values['class'],
    ...values.props,
  },
  m('div.outside', null,
    m('div.inside', null, values.child)))
//##
            );
        }

        return (
//##
m(values.tag, {'class': values['class'], ...values.props}, values.child)
//##
        );
    }
}


export class AspectRatio extends Node
{
    WANT_CHILDREN = true
    NODE_PROPS = ['ratio']

    prepare(vnode, values, context)
    {
        let ratio = vnode.attrs.ratio;
        if (ratio)
        {
            values['class'].push(`bx--aspect-ratio--${ratio}`);
        }
    }

    render_default(vnode, values, context)
    {
        return (
//##
m(values.tag,
  {
    'class': `bx--aspect-ratio ${values['class']}`,
    ...values.props,
  },
  m('div.bx--aspect-ratio--object', null, values.child))
//##
        );
    }
}
